### PR TITLE
avm2: Partially implement Stage3D for wgpu backend

### PR DESCRIPTION
**Pull request by Aaron1011**
This PR implements core 'stage3D' APIs. We are now able
    to render at least two demos from the Context3D docs - a simple
    triangle render, and a rotating cube.
    
Implemented in this PR:
    * Stage3D access and Context3D creation
    * IndexBuffer3D and VertexBuffer3D creation, uploading, and usage
    * Program3D uploading and usage (via `naga-agal`)
    * Context3D: configureBackBuffer, clear, drawTriangles, and present
    
Not yet implemented:
    * Any 'dispose()' methods
    * Depth and stencil buffers
    * Context3D texture apis
    * Scissor rectangle
    
General implementation strategy:
    
A new `Object` variant is added for each of the Stage3D objects
    (VertexBuffer3D, Program3D, etc). This stores a handle to the
    parent `Context3D`, and (depending on the object) a handle
    to the underlying native resource, via `Rc<dyn
    SomeRenderBackendTrait>`).
    
Calling methods on Context3D does not usually result in an immediate
    call to a `wgpu` method. Instead, we queue up commands in our
    `Context3D` instance, and execute them all on a call to `present`.
    This avoids some nasty wgpu lifetime issues, and is very similar
    to the approah we use for normal rendering.
    
The actual rendering happens on a `Texture`, with dimensions
    determined by `createBackBuffer`. During 'Stage' rendering,
    we render all of these Stage3D textures *behind* the normal
    stage (but in front of the overall stage background color).

### CODE REVIEW

File: core/build_playerglobal/src/lib.rs [hunk 1]
```diff
@@ -120,10 +120,22 @@ fn flash_to_rust_path(path: &str) -> String {
...
+            }
+
+            let mut without_boundaries = vec![Boundary::DigitUpper];
+            // Special case for classes ending in '3D' - we want to ave something like
+            // 'vertex_buffer_3d' instead of 'vertex_buffer3d'
+            if !component.ends_with("3D") {
+                // Do not split on a letter followed by a digit, so e.g. `atan2` won't become `atan_2`.
+                without_boundaries.extend(&[Boundary::UpperDigit, Boundary::LowerDigit]);
+            }
+
```
Comment (EmperorBale) [inline core/build_playerglobal/src/lib.rs:135]: I think there should maybe be a better way of doing this, but we can do that in a different PR. An idea I was thinking of was setting the rust name in metadata in actionscript, like:
```as
[rust_name("vertex_buffer_3d")]
public class VertexBuffer3D {}
```

File: core/src/avm2.rs [hunk 2]
```diff
@@ -116,7 +116,7 @@ impl<'gc> Avm2<'gc> {
             broadcast_list: Default::default(),
 
             #[cfg(feature = "avm_debug")]
-            debug_output: false,
+            debug_output: true,
```
Comment (EmperorBale) [inline core/src/avm2.rs:119]: I don't think you meant to commit this

**Comment by torokati44** [comment]
The included `Matrix3D` class in itself (I think) already fixes a handful of z0r loops! \o/

**Comment by Aaron1011** [comment]
This is now ready for review

**Comment by kmeisthax** [comment]
@Aaron1011 I'm about a third of the way through reading this, but I noticed that this PR switches dev from abort-panics to unwinding. Is there any particular feature of this PR that requires that to be set?

**Comment by Aaron1011** [comment]
@kmeisthax I committed that by accident - it can be reverted

**Review by kmeisthax — APPROVED** [review]
I'm assuming this is ready to merge. The foundations look pretty good - the only thing I'm a little iffy about is the comment you made about strict wgpu lifetimes. I suspect that's unavoidable complexity, however.

The PR's been sitting in the queue for a while and has some conflicts you'll need to resolve first. If you can get those done, I can get it merged tonight or tomorrow.

File: core/src/avm2/globals/flash/display3D/index_buffer_3d.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+) -> Result<Object<'gc>, Error<'gc>> {
+    panic!("This allocator should not be called!")
+}
+
+pub fn upload_from_byte_array<'gc>(
+    _activation: &mut Activation<'_, 'gc, '_>,
+    _this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    Ok(Value::Undefined)
```
[the remark this answers is not in the document]
Comment (Aaron1011) [reply core/src/avm2/globals/flash/display3D/index_buffer_3d.rs:19]: I've added a `log::warn!`
